filters/openpolicyagent/authorizewithregopolicy/authorizewithregopolicy.go

```diff
@@ -0,0 +1,160 @@
+	}
+
+	bundleName := sargs[0]
+
+	configOptions := s.opts
+
+	envoyContextExtensions := map[string]string{}
+	if len(sargs) > 1 {
+		err = yaml.Unmarshal([]byte(sargs[1]), &envoyContextExtensions)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	opaConfig, err := openpolicyagent.NewOpenPolicyAgentConfig(configOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	opa, err := s.registry.NewOpenPolicyAgentInstance(bundleName, *opaConfig, s.Name())
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &authorizeWithRegoPolicyFilter{
+		opa:                    opa,
+		registry:               s.registry,
+		envoyContextExtensions: envoyContextExtensions,
+	}, nil
+}
+
+type authorizeWithRegoPolicyFilter struct {
+	opa                    *openpolicyagent.OpenPolicyAgentInstance
+	registry               *openpolicyagent.OpenPolicyAgentRegistry
+	envoyContextExtensions map[string]string
+}
+
+func (f *authorizeWithRegoPolicyFilter) Request(fc filters.FilterContext) {
+	req := fc.Request()
+	span, ctx := f.opa.StartSpanFromFilterContext(fc)
+	defer span.Finish()
+
+	authzreq := envoy.AdaptToExtAuthRequest(req, f.opa.InstanceConfig().GetEnvoyMetadata(), f.envoyContextExtensions)
+
+	start := time.Now()
+	result, err := f.opa.Eval(ctx, authzreq)
+	fc.Metrics().MeasureSince(f.opa.MetricsKey("eval_time"), start)
+
+	if err != nil {
+		f.opa.RejectInvalidDecisionError(fc, span, result, err)
+		return
+	}
+
+	if f.opa.EnvoyPluginConfig().DryRun {
+		return
+	}
+
+	allowed, err := result.IsAllowed()
+
+	if err != nil {
+		f.opa.RejectInvalidDecisionError(fc, span, result, err)
+		return
+	}
+
+	if !allowed {
+		fc.Metrics().IncCounter(f.opa.MetricsKey("decision.deny"))
+		f.opa.ServeResponse(fc, span, result)
+		return
+	}
+
+	fc.Metrics().IncCounter(f.opa.MetricsKey("decision.allow"))
+
+	if result.HasResponseBody() {
+		body, err := result.GetResponseBody()
+		if err != nil {
+			f.opa.RejectInvalidDecisionError(fc, span, result, err)
+			return
+		}
+		fc.StateBag()[openpolicyagent.OpenPolicyAgentDecisionBodyKey] = body
+	}
+
+	headers, err := result.GetResponseHTTPHeaders()
+	if err != nil {
+		f.opa.RejectInvalidDecisionError(fc, span, result, err)
+		return
+	}
+	addRequestHeaders(fc, headers)
+
+	headersToRemove, err := result.GetRequestHTTPHeadersToRemove()
+	if err != nil {
+		f.opa.RejectInvalidDecisionError(fc, span, result, err)
+		return
+	}
+	removeHeaders(fc, headersToRemove)
+}
+
+func removeHeaders(fc filters.FilterContext, headersToRemove []string) {
+	for _, header := range headersToRemove {
+		fc.Request().Header.Del(header)
+	}
+}
+
+func addRequestHeaders(fc filters.FilterContext, headers http.Header) {
+	for key, values := range headers {
+		for _, value := range values {
+			fc.Request().Header.Add(key, value)
+		}
+	}
+}
+
+func (*authorizeWithRegoPolicyFilter) Response(filters.FilterContext) {}
+
+func (f *authorizeWithRegoPolicyFilter) Close() error {
+	return f.registry.ReleaseInstance(f.opa)
+}
```

Comment:
This likely does not work as you'd expect. Basically it is never called, see https://github.com/zalando/skipper/issues/202

Reply:
I had a look and I am a bit puzzled. When running this code, I could see that the instances were indeed cleaned up (in the control plane you otherwise see dormant / dead instances)

Reply:
Ok, so instances were properly re-used which is why it looked like a clean up. As we never removed a specific instance, we also did not observe missing close. Changed this to a PostProcessor to detect in-active instances